core/src/main/java/io/grpc/ManagedChannelProvider.java

```diff
@@ -87,6 +87,27 @@ public static ManagedChannelProvider provider() {
     return provider;
   }
 
+  private static ClassLoader getCorrectClassLoader() {
+    if (isAndroid()) {
+      // When android:sharedUserId or android:process is used, Android will setup a dummy
+      // ClassLoader for the thread context (http://stackoverflow.com/questions/13407006),
+      // instead of letting users to manually set context class loader, we choose the
+      // correct class loader here.
+      return ManagedChannelProvider.class.getClassLoader();
```

Comment:
`Class.forName` would use `ManagedChannelProvider.class.getClassLoader()`, to my understanding.

There are some concerns about using the write class loader _in general_, but on Android @madongfly was saying that all classes in the same APK are in the same class loader, which is what I would have expected.